File: cpp/template_method/template_method.cpp
#include <iostream>
#include <memory>

class CaffeineBeverage {
protected:
    void _boil_water() {
        std::cout << "Boiling Water!" << std::endl;
    }

    virtual void _add_components() {
        throw std::logic_error("Not implemented error.");
    }

public:
    virtual void prepare() {
        _boil_water();
        _add_components();
    }
};

class Tea : public CaffeineBeverage {
protected:
    void _add_components() override {
        std::cout << "Adding Lemons!" << std::endl;
    }
};

int main() {
    Tea tea;
    tea.prepare();
    return 0;
}

File: python/template_method/template_method.py
from abc import ABC


class CaffeineBeverage(ABC):
    def _boil_water(self):
        print("Boiling Water!")

    def _add_components(self):
        raise NotImplementedError()

    def prepare(self):
        self._boil_water()
        self._add_components()


class Tea(CaffeineBeverage):
    def _add_components(self):
        print("Adding Lemons!")


def main():
    tea = Tea()
    tea.prepare()


if __name__ == '__main__':
    main()
